src/librustc_mir/interpret/eval_context.rs

```diff
@@ -136,7 +136,8 @@ impl<'tcx, Tag: Copy + 'static> LocalState<'tcx, Tag> {
         match self.value {
             LocalValue::Dead => throw_unsup!(DeadLocal),
             LocalValue::Uninitialized =>
-                bug!("The type checker should prevent reading from a never-written local"),
+                // this is reachable from ConstProp
+                throw_unsup!(UninitializedLocal),
```

Comment:
Is this ever caught? Any reason not to use `throw_unsup_format`?

Actually it seems to me this is a case for `throw_inval`; after all the only reason this is reachable is that const-prop runs on non-type-checked programs. Which, btw, is quite annoying... what is the point of doing all this work on code that will fail compilation anyway? The user pays by having to wait longer until compilation is done, and we pay by having tons of "should be unreachable but const-prop" branches in the Miri engine, and everyone is worse off.

Reply:
>  after all the only reason this is reachable is that const-prop runs on non-type-checked programs. Which, btw, is quite annoying... what is the point of doing all this work on code that will fail compilation anyway?

Does const-prop run on non-type-checked programs? It's one of the very last passes to run in the `optimized_mir` query so it shouldn't be running on code which is going to fail compilation.

https://github.com/rust-lang/rust/blob/b6269f27d99d7da9e95f0b3fdc53193dc8c42fbe/src/librustc_mir/transform/mod.rs#L267

Reply:
Well but then how can it hit this code path here, and well-typed normal code run by Miri cannot?

Reply:
ConstProp doesn't currently allow propagation between locals

https://github.com/rust-lang/rust/blob/a44881d892fb4f4a8ed93f8f392bab942fac7a41/src/librustc_mir/transform/const_prop.rs#L603-L607

So for something like:

```rust
let x = 4; // _1
let y = x; // _2
```

https://github.com/rust-lang/rust/blob/a44881d892fb4f4a8ed93f8f392bab942fac7a41/src/librustc_mir/transform/const_prop.rs#L679

`self.can_const_prop[_1]` will be `false` so we won't store `4` to `_1`. As a result, when we eval `_2 = _1`, `_1` will be uninitialized.

Reply:
Oh I see, so the problem is that ConstProp "forgets" the value stored in locals. The comment here should say that.

It is still a bug to ever reach this during normal execution, so this feels more like a case for `throw_inval` to me.

Reply:
> It is still a bug to ever reach this during normal execution, so this feels more like a case for throw_inval to me.

Hm, I cannot make up my mind. Maybe invalid is actually okay as this is comparable to ptr-to-int casts during CTFE... they are supported in Miri, but not supported in other instances of the engine. But I also want to be sure that we ICE if we ever hit this in Miri (and eventually Miri will ICE when hitting an "invalid program" error). Any suggestions for how to achieve this?

Reply:
We can create a separate machine for const prop and add a hook for this

Reply:
I am leaning more and more towards a dedicated const-prop machine as well, yes.